### ISSUE TITLE
Implementing a high quality import/export feature 

### ISSUE THREAD

**Issue by MLI-DS** [django-oscar/django-oscar#3498]
### Issue Summary

I am looking to implement a import export feature in my system and figured this could be a feature that could benefit Oscar as a whole,  it seems to be a basic feature in other E-commerce platforms / OMS.

Based of the documentation of importing products here: https://django-oscar.readthedocs.io/en/stable/howto/importin,g_a_catalogue.html

I would make a new I/O app( unsure wether this is a good idea, but  the current importer is under oscar.apps.partner.importers which seems a little odd) 
The importer would be based on the same logic described in the documentation ( parse csv to Object)
The exporter would be based on the oposite logic (write Object to csv)
The only format supported would be csv at first but this can be further extended
I would then add a view + template to import/export the catalogue ( in its integrity at first, we can imagine extending this to export a particular subset of the catalogue or a model other then products)

As for the file structure I would go

- impex
  - utils
    - parsers
       - csv.py
    - writers
       - csv.py 
  - views.py
  - apps.py
 
I don't think any new models would be necessary for this app.

I would then link the views to import export to a new tab on the dashboard
.
Please let me know this solution is good, if I am missing something or am not approaching it the right way.

**Comment by solarissmoke** [comment]
Hey @MLI-DS thanks for this - I think it would make a very useful addition to the core. The approach and app structure you've proposed looks good to me - would be great if you have time to work on a PR for this! A couple of things that come to mind:

1. While dashboard views are useful, the parsers/writers should be standalone and usable in other contexts (management commands, tasks etc) like the current `CatalogueImporter` is.

2. It would be nice to find a generic way to support import/export of product attributes, bearing in mind that these can vary by product class. I've not spent much time looking at what other frameworks do but I'm sure there are ways to address this that we can look at.

**Comment by MLI-DS** [comment]
Thanks for the pointers :+1:  

1. As for making the parsers/writers standalone that is what I had in mind too, and placed them under a utils to be used in other contexts if need be. Did you have something more specific in mind ? 

2. I was going to inspire myself of the oscarapi serializers for the objects and then use something like [django-rest-framework-csv]( https://github.com/mjumbewu/django-rest-framework-csv )to get the csv output (potentially coded internally to avoid relying third party code).

 Does this seem good to you ? What partiuclar difficulties could be faced when importing/exporting product attributes?

**Comment by solarissmoke** [comment]
>  Did you have something more specific in mind?

No, I think what you've proposed works, just wanted to state it explicitly.

> What particular difficulties could be faced when importing/exporting product attributes?

The main issue is that these vary by product class, which makes it awkward to fit them into CSV columns when exporting products of more than one class. Imagine that products of class `Books` have attributes `isbn`, `author` and `num_pages`, and products of class `Clothing` have attributes `size` and `colour` - we need to be able to export products of either class, and their attributes. This is where a format like CSV runs into limitations that something like JSON would not.

The simplest option would simply be to only allow import/export of one product class at a time (for CSV). A more complex option would require coming up with an aggregate set of columns for all possible product classes. I don't have an opinion yet on what the right approach is - would want to investigate a bit more how other frameworks handle this.

**Comment by rodrigoegimenez** [comment]
Hi! I was just thinking about starting to work on a similar module. I have already written an importer and wanted to give back some of the code, making it more generic and useful.

I had the idea of creating an importer model to be able to keep track of imports and also do them asynchronously (submit the file, create and save the form, and trigger the import function). Reading your thoughts a good implementation would be to perhaps to use JSON as a base format for import-export and then have helper functions to convert  this JSON to/from CSV and sheet document files (using the open format ODS). In this case a good file structure might be: one CSV file per product class and in the ODS file one sheet per product class.

**Comment by MLI-DS** [comment]
I though of using models for importers and exporters to be able to save the metadata and potentially replay imports and exports, I refrained from adding it in here as I though this might make the feature heavier then necessary ?

Using JSON as a base seems like a good solution!, Its widely used and as @solarissmoke pointed out bypasses the limitations of CSV when working with product classes. 

Allowing IO to/from ODS format seems quite elegant for end users, as they avoid having lots of different files to store their catalogue and ends up being a simple solution (which is bound to happen if we only implement csv). 

To benchmark with Woocommerce import/export system ( which uses csv), each product attribute has 2 columns in the product row one for the name and one for the value, this could also be a good solution.
e.g: 

name Attribute 1 | Value attribute 1 | name Attribute 2 | Value Attribute 2
isbn | 123458382 | author | JCVD ...
isbn | 09865341 | author | Jet Li

**Comment by meigallodixital** [comment]
Why not use as base https://github.com/django-import-export/django-import-export, or https://github.com/jazzband/tablib if we need to be as agnostic as possible?
